### PR TITLE
Use LXC for development environment

### PR DESCRIPTION
**Pull request by danypr92**
I propose this development playbook and this automation scripts to help create a development environment using [LXC](https://linuxcontainers.org/) in your local Linux machine.

This proposal is **only for the development environment**, it doesn't affect how we provision production and staging machines. 

The idea is to use `default_user.yml` playbook to create an admin user and use `development.yml` playbook to provision the container with the tools needed to run the application.

We create the following Quick Start guide to help new user to install the development environment and to make it easy for the new contributors.

# OFN Development Environment
## Quick start

This document will guide you through creating an LXC container which you start/stop to develop ofn,
 provision it with everything you need to develop ofn and provide you with easy commands to get up and running with ofn as quickly and as painlessly as possible.

### Who is this document for?

Anyone, wishing to contribute to the Open Food Foundation project and is using Linux and can run [LXC](https://linuxcontainers.org/) on their host machine.

### Step 1 - Checkout code and configure Ansible
    
  Checkout the following two repos in the same folder:

    git clone git@github.com:openfoodfoundation/ofn-install.git
    git clone git@github.com:openfoodfoundation/openfoodnetwork.git

  Copy `ofn-install/inventory/host_vars/local.ofn.org/secrets.example.yml` to `ofn-admin/inventory/host_vars/local.ofn.org/secrets.yml`

    # Keep this password as is (same as that in database.yml in openfoodnetwork app)!!
    db_password: f00d
    admin_password: f00d
 
    # Secret Token
    # This is required to verify session. It must be at least 30 characters and
    # random.
    # Generate one with a bash command like: openssl rand -hex 128
    secret_token: 'xxxxxx'

### Step 2 - Configuration file
```
# scripts/config/lxc.cfg

# Vars to use in lxc-create
name="ofn-dev"
template="/usr/share/lxc/templates/lxc-ubuntu"
rls="xenial"
lxc_config="/tmp/ubuntu.$name.conf"
host="local.ofn.org"
project_name="openfoodnetwork"
project_path="${PWD%/*}/$project_name"
app_user="openfoodnetwork"
ssh_path="/.ssh/id_rsa.pub"
```
Modify `scripts/config/lxc.cfg` to change the container name, host name or path of your ssh key.

### Step 3 - Create LXC container

    cd ofn-install
    ./scripts/create-container.sh

  If this has worked correctly, you shouldn't have any trouble doing **both** of the following:

    ssh openfoodnetwork@local.ofn.org
    ssh root@local.ofn.org 

  It should not require passwords, because your ssh key should already be copied to the container.

### Step 4 - Provision LXC container
  Ensure you have the lastest (>=2.3) Ansible package installed. 
  For instance in Ubuntu would be:
  
    sudo apt-add-repository ppa:ansible/ansible
    sudo apt update
    sudo apt install ansible

  You can find more information about Ansible installation in the [Ansible Documentation Site](http://docs.ansible.com/ansible/latest/intro_installation.html) 

  Then you can provision the LXC container:

    ./scripts/provisioning-dev.sh

  *Note:* The first **SUDO** password it asks for is your host user password.

### Step 5 - Start Rails server

  Login in your container via ssh `ssh openfoodnetwork@local.ofn.org`, go to the project folder `cd openfoodnetwork` and run `bundle exec rails server` or `bundle exec rails s` to start the server.

  Now you can visit `local.ofn.org:3000` and look your local development instance.

### CODE REVIEW

File: inventory/host_vars/local.ofn.org/config.yml [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+
+domain: local.ofn.org
+rails_env: development
+admin_email: admin@example.com
```
Comment (enricostano) [inline inventory/host_vars/local.ofn.org/config.yml]: is `admin_email` used in development?

File: inventory/host_vars/local.ofn.org/config.yml [hunk 2]
```diff
@@ -0,0 +1,5 @@
+---
+
```
Comment (enricostano) [inline inventory/host_vars/local.ofn.org/config.yml]: Remove space

File: playbooks/development.yml [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+      become_user: "{{ unicorn_user }}"
+      tags: bundler
+
+    - role: dbserver # Set up database server and user for the app.
+      tags: dbserver
+      db_user_roles: SUPERUSER,CREATEDB
+      db_user: ofn
+
+    # - role: devloper # Handy tools for developers
+    #   tags: devloper
```
Comment (enricostano) [inline playbooks/development.yml]: What is this?

File: scripts/config/lxc.cfg [hunk 4]
```diff
@@ -0,0 +1,10 @@
+# Vars to use in lxc-create
+name=ofn-dev
+template=/usr/share/lxc/templates/lxc-ubuntu
+rls=trusty
```
Comment (enricostano) [inline scripts/config/lxc.cfg]: `xenial` at least?

File: scripts/config/lxc.cfg [hunk 5]
```diff
@@ -0,0 +1,10 @@
...
+# config="/tmp/ubuntu.conf"
+host=local.ofn.org
+nproject=openfoodnetwork
```
Comment (enricostano) [inline scripts/config/lxc.cfg]: `project_name` would be easier to understand IMHO

File: scripts/config/lxc.cfg [hunk 6]
```diff
@@ -0,0 +1,10 @@
...
+fproject="${PWD%/*}/$nproject"
```
Comment (enricostano) [inline scripts/config/lxc.cfg]: Same here

File: scripts/create-container.sh [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+
+# Default values
+name="ofn-test"
+template="/usr/share/lxc/templates/lxc-ubuntu"
+rls="trusty"
+dconfig="/tmp/ubuntu.$name.conf"
+host="ofn-test.org"
+nproject="openfoodnetwork"
+fproject="${PWD%/*}/$nproject"
+app_user="openfoodnetwork"
```
Comment (enricostano) [inline scripts/create-container.sh]: Do we need these variables defined here? We'll get them with the next `source` right?

**Comment by danypr92** [comment]
Thanks @enricostano!

**Comment by mkllnk** [comment]
I prefer to run OFN directly on my OS (still using Debian 7). But this could be a great option for people who like to separate environments and have a machine that is a bit more up-to-date than mine.

**Comment by enricostano** [comment]
@mkllnk exactly, I use Arch Linux on my machine and I like to have an LXC container for each project I work on. So I don't have to deal with crossed dependencies and I'm always sure that each project uses the same dependencies as in production/staging. But that doesn't mean that everybody must use this of course.

**Pull request merged**